weave/trace/op.py

```diff
@@ -734,4 +785,294 @@ def as_op(fn: Callable) -> Op:
+            self._iterator = iter(self._context_value)
+            return self
+        return self._context_value
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        if self.op.lifecycle_handler.has_finished:
+            raise OpCallError("Should not call finish more than once")
+
+        boxed_output = box.box(self.call.output)
+        client = weave_client_context.require_weave_client()
+        client.finish_call(self.call, boxed_output, exc_value, op=self.op)
+        # self.op.lifecycle_handler.has_finished = True
+
+        if not call_context.get_current_call():
+            print_call_link(self.call)
+        call_context.pop_call(self.call.id)
+
+        return self.orig_contextmanager.__exit__(exc_type, exc_value, traceback)
+
+
+class AsyncIterableContext:
+    """Wraps an op that returns an async iterable contextmanager to add relevant lifecycle hooks.
+
+    This is used for streaming ops, where we want to accumulate the output of the op until
+    the context manager is closed.  This is currently only used for Anthropic streaming ops.
+    """
+
+    def __init__(self, op: Op, args: Any, kwargs: Any, call: Call, should_raise: bool):
+        self.op = op
+        self.args = args
+        self.kwargs = kwargs
+        self.call = call
+        self.should_raise = should_raise
+        self.orig_contextmanager = op.resolve_fn(*args, **kwargs)
+
+        # If the context manager is also iterable...
+        self._context_value = None
+        self._aiterator = None
+
+    async def __aenter__(self):
+        self._context_value = await self.orig_contextmanager.__aenter__()
+        if hasattr(self._context_value, "__aiter__"):
+            self._aiterator = self._context_value.__aiter__()
+            return self
+        return self._context_value
+
+    async def __aexit__(self, exc_type, exc_value, traceback):
+        if self.op.lifecycle_handler.has_finished:
+            raise OpCallError("Should not call finish more than once")
+
+        boxed_output = box.box(self.call.output)
+        client = weave_client_context.require_weave_client()
+        client.finish_call(self.call, boxed_output, exc_value, op=self.op)
+        # self.op.lifecycle_handler.has_finished = True
+
+        if not call_context.get_current_call():
+            print_call_link(self.call)
+        call_context.pop_call(self.call.id)
+
+        return await self.orig_contextmanager.__aexit__(exc_type, exc_value, traceback)
+
+    def __aiter__(self):
+        return self
+
+    async def __anext__(self):
+        if self._aiterator is None:
+            raise TypeError("Context value is not async iterable")
+        try:
+            val = await self._aiterator.__anext__()
+        except StopAsyncIteration:
+            raise
+        self.op.lifecycle_handler.before_yield(self.call, val)
+        return val
+
+
+@contextmanager
+def _call_context(op: Op, call: Call, should_raise: bool):
```

Comment:
All op execution happens in this context.  This block basically says:
1. Do your thing
2. If you fail, run the error callback
3. Always try to finish the call out